rasa_nlu/server.py

```diff
@@ -48,74 +47,109 @@ def requires_auth(f):
         else:
-            request_params = request.get_json(force=True)
+            request_params = json.loads(request.content.read().decode('utf-8', 'strict'))
         if 'q' not in request_params:
-            return jsonify(error="Invalid parse parameter specified"), 404
+            request.setResponseCode(404)
+            returnValue(json.dumps({"error": "Invalid parse parameter specified"}))
         else:
+            data = self.data_router.extract(request_params)
             try:
-                data = current_app.data_router.extract(request_params)
-                response = current_app.data_router.parse(data)
-                return jsonify(response)
+                request.setResponseCode(200)
+                response = yield (maybeDeferred(self.data_router.parse, data)
+                                  if self.testing
+                                  else threads.deferToThread(self.data_router.parse, data))
+                returnValue(json.dumps(response))
             except InvalidModelError as e:
-                return jsonify({"error": "{}".format(e)}), 404
+                request.setResponseCode(404)
+                returnValue(json.dumps({"error": "{}".format(e)}))
 
-    @rasa_nlu_app.route("/version", methods=['GET'])
+    @app.route("/version", methods=['GET'])
     @requires_auth
-    def version():
-        return jsonify({'version': __version__})
+    def version(self, request):
+        """Returns the Rasa server's version"""
 
-    @rasa_nlu_app.route("/config", methods=['GET'])
-    @requires_auth
-    def rasaconfig():
-        return jsonify(config.as_dict())
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps({'version': __version__})
 
-    @rasa_nlu_app.route("/status", methods=['GET'])
+    @app.route("/config", methods=['GET'])
     @requires_auth
-    def status():
-        return jsonify(current_app.data_router.get_status())
+    def rasaconfig(self, request):
+        """Returns the in-memory configuration of the Rasa server"""
 
-    @rasa_nlu_app.route("/", methods=['GET'])
-    def hello():
-        return "hello from Rasa NLU: " + __version__
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps(self.config.as_dict())
 
-    @rasa_nlu_app.route("/train", methods=['POST'])
+    @app.route("/status", methods=['GET'])
     @requires_auth
-    def train():
-        data_string = request.get_data(as_text=True)
-        current_app.data_router.start_train_process(data_string, request.args)
-        return jsonify(info="training started.", training_process_ids=current_app.data_router.train_proc_ids())
+    def status(self, request):
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps(self.data_router.get_status())
+
+    @app.route("/train", methods=['POST'])
+    @requires_auth
+    def train(self, request):
+        def errback(f):
```

Comment:
but make sure to mention that in the changelog and the migration file (as anyone calling that api will now need to wait till it is finished training)